scripts/install_staging.sh

```diff
@@ -0,0 +1,125 @@
+#!/bin/bash
+
+# Update your system packages
+install_mac() {
+    # Ensure Homebrew is installed
+    which brew > /dev/null
+    if [ $? -ne 0 ]; then
+        echo "Installing Homebrew..."
+        /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
+    fi
+
+    echo "Updating Homebrew packages..."
+    brew update
+
+    echo "Installing required packages..."
+    brew install make llvm curl libssl protobuf tmux
+}
+
+# Function to install packages on Ubuntu/Debian
+install_ubuntu() {
+    echo "Updating system packages..."
+    sudo apt update
+
+    echo "Installing required packages..."
+    sudo apt install --assume-yes make build-essential git clang curl libssl-dev llvm libudev-dev protobuf-compiler tmux
+}
+
+# Detect OS and call the appropriate function
+if [[ "$OSTYPE" == "darwin"* ]]; then
+    install_mac
+elif [[ "$OSTYPE" == "linux-gnu"* ]]; then
+    install_ubuntu
+else
+    echo "Unsupported operating system."
+fi
+
```

Comment:
We should thinkg carefully about the following:
1. First time running the script 
2. All additional times (reinstall / create wallet 

I think it's a better design to have two scripts (or at least logical sections that only run when necessary):
1. Build/install (OS requirements, libraries, btcli interactions)
2. Test/run (only python scripts)

Reply:
How about this?

```bash
#!/bin/bash

# Section 1: Build/Install
# This section is for first-time setup and installations.

install_dependencies() {
    # Function to install packages on macOS
    install_mac() {
        which brew > /dev/null
        if [ $? -ne 0 ]; then
            echo "Installing Homebrew..."
            /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
        fi
        echo "Updating Homebrew packages..."
        brew update
        echo "Installing required packages..."
        brew install make llvm curl libssl protobuf tmux
    }

    # Function to install packages on Ubuntu/Debian
    install_ubuntu() {
        echo "Updating system packages..."
        sudo apt update
        echo "Installing required packages..."
        sudo apt install --assume-yes make build-essential git clang curl libssl-dev llvm libudev-dev protobuf-compiler tmux
    }

    # Detect OS and call the appropriate function
    if [[ "$OSTYPE" == "darwin"* ]]; then
        install_mac
    elif [[ "$OSTYPE" == "linux-gnu"* ]]; then
        install_ubuntu
    else
        echo "Unsupported operating system."
        exit 1
    fi

    # Install rust and cargo
    curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh

    # Update your shell's source to include Cargo's path
    source "$HOME/.cargo/env"
}

# Call install_dependencies only if it's the first time running the script
if [ ! -f ".dependencies_installed" ]; then
    install_dependencies
    touch .dependencies_installed
fi

# Section 2: Test/Run
# This section is for running and testing the setup.

# Logic for setting up and running the environment
setup_environment() {
    # Clone subtensor and enter the directory
    if [ ! -d "subtensor" ]; then
        git clone https://github.com/opentensor/subtensor.git
    fi
    cd subtensor
    git pull

    # Update to the nightly version of rust
    ./scripts/init.sh

    cd ../bittensor-subnet-template

    # Install the bittensor-subnet-template python package
    python -m pip install -e .

    # Create and set up wallets
    # This section can be skipped if wallets are already set up
    if [ ! -f ".wallets_setup" ]; then
        btcli wallet new_coldkey --wallet.name owner --no_password --no_prompt
        btcli wallet new_coldkey --wallet.name miner --no_password --no_prompt
        btcli wallet new_hotkey --wallet.name miner --wallet.hotkey default --no_prompt
        btcli wallet new_coldkey --wallet.name validator --no_password --no_prompt
        btcli wallet new_hotkey --wallet.name validator --wallet.hotkey default --no_prompt
        touch .wallets_setup
    fi

    # Additional script logic goes here...
}

# Call setup_environment every time
setup_environment

# Additional logic for running and testing the setup goes here...

```

Reply:
This way we can have a touch file to denote if we need to resetup wallets and whatnot.